
# Data Analytics Test
## Submitted By: NEHA
### College: Laxmi Devi Institute of Engineering and Technology, Alwar
### Date: 18-02-2026

# Q1. What will be the output of the following Python code?


In [1]:
def func(a, b=5, c=10):
    return a + b * c

print(func(2, c=3))

17


EXPLANATION:
In this function:

a = 2 (first positional argument)

b = 5 (default value, since it wasn’t provided)

c = 3 (overridden using keyword argument)

Now calculate:

a + b * c
= 2 + 5 * 3
= 2 + 15
= 17


Multiplication happens before addition (operator precedence), so the final output is:

17

# Q2. What is the scope of a variable declared inside a function?

In [2]:
def my_func():
    x = 10   # local variable
    print("Inside function:", x)

my_func()

# print(x)  # This will give an error because x is not accessible outside

Inside function: 10


EXPLANATION:
* The variable x is created inside the function, so it has local scope.

* Local variables exist only within the function where they are defined.

When my_func() runs, it prints:

Inside function: 10


After the function finishes execution, the local variable x is destroyed.

If you uncomment print(x) outside the function, Python will raise a NameError because x is not defined in the global scope.

So, x is accessible only inside my_func(), not outside it.

# Q3. What does the following code return?

In [3]:
def f(x, lst=[]):
    lst.append(x)
    return lst

print(f(1))
print(f(2))

[1]
[1, 2]


EXPLANATION:
* The key concept here is mutable default arguments.

* The default value lst=[] is created only once, when the function is defined — not each time the function is called.

Since lists are mutable, changes to lst persist between function calls.

Step-by-step execution:

1️⃣ First call: f(1)

lst is the default empty list []

1 is appended → [1]

Output:

[1]


2️⃣ Second call: f(2)

The same list is reused (it is not reset)

Current list is [1]

2 is appended → [1, 2]

Output:

[1, 2]

Final Output:
[1]
[1, 2]


This happens because the default list is shared across function calls.

# 4. Which statement about *args is correct?

In [4]:
def add_numbers(*args):
    return sum(args)

print(add_numbers(1, 2, 3, 4))


10


EXPLANATION:
*args

The *args syntax allows the function to accept a variable number of positional arguments.

All arguments passed are collected into a tuple named args.

In this call: add_numbers(1, 2, 3, 4), args becomes (1, 2, 3, 4).

sum(args)

The built-in sum() function adds all numbers in an iterable.

Here, sum((1, 2, 3, 4)) = 10.

Return value

The function returns 10.

print()

Finally, print() outputs the returned value:

10

# 5. What will be the output of the following code?

In [5]:
x = 10

def test():
    x = 5
    return x

print(test(), x)

5 10


EXPLANATION:
There are two variables named x:

1️⃣ Global variable → x = 10  
2️⃣ Local variable inside function → x = 5  

When test() is called:
- It returns the local value of x, which is 5.

Outside the function:
- The global x remains unchanged (10).

# 6. Which keyword is used to modify a global variable inside a function?

In [6]:
x = 10

def change_value():
    global x
    x = 20

change_value()
print(x)

20


EXPLANATION:

Global variable x

x = 10 is defined in the global scope.

global x inside the function

The statement global x tells Python that the x inside the function refers to the global variable, not a new local variable.

Changing the value

x = 20 now modifies the global x, because of the global keyword.

Calling the function

change_value() runs, and the global x is updated to 20.

Printing x

print(x) outputs the updated global value:

Output:
20

# 7. What is returned if a function has no return statement?

In [7]:
def my_function():
    x = 10  # no return statement

result = my_function()
print(result)


None


EXPLANATION:
In Python, if a function does not have a return statement,
it automatically returns **None**.

None is a special value that represents the absence of a value.


# 8. Which statement is true about lambda functions?

In [8]:
square = lambda x: x * x
print(square(5))



25


EXPLANATION:
lambda x: x * x creates a one-line function to square a number.

square(5) → 5 * 5 = 25.

Output:

25

# Q9. What is the primary advantage of NumPy arrays over Python lists?

In [9]:
import numpy as np

arr = np.array([1, 2, 3])
print(arr * 2)


[2 4 6]


EXPLANATION:
* arr is a NumPy array [1, 2, 3].

* Multiplying by 2 scales each element of the array: [1*2, 2*2, 3*2] = [2, 4, 6].

# 10. What will be the output of the following NumPy code?

In [10]:
import numpy as np

a = np.array([1, 2, 3])
print(a * 2)


[2 4 6]


EXPLANATION:
* np.array([1, 2, 3])

Creates a NumPy array a with elements [1, 2, 3].

Element-wise multiplication

In NumPy, multiplying an array by a number applies to each element individually.

So a * 2 becomes [1*2, 2*2, 3*2] = [2, 4, 6].

Printing the result

print(a * 2) outputs:

[2 4 6]

# 11. What does broadcasting allow in NumPy?

In [11]:
import numpy as np

a = np.array([1, 2, 3])
b = 5

print(a + b)

[6 7 8]


EXPLANATION:
* a is a NumPy array [1, 2, 3].

* b is a scalar 5.

* Broadcasting in NumPy

When a scalar is added to an array, NumPy adds the scalar to each element of the array automatically.

So a + b = [1+5, 2+5, 3+5] = [6, 7, 8].

Printing the result

Output:

[6 7 8]

# Question 12. What is the shape of the array created by:

## np.zeros((3,4))


In [12]:
import numpy as np

a = np.zeros((3, 4))
print(a.shape)


(3, 4)


EXPLANATION:
np.zeros((3, 4)) creates a 3×4 array filled with zeros.

.shape returns the dimensions of the array.

Output:

(3, 4)

# 13. What does np.arange(0, 10, 2) return?

In [13]:
import numpy as np

arr = np.arange(0, 10, 2)
print(arr)


[0 2 4 6 8]


EXPLANATION:
np.arange(start, stop, step) creates an array starting at start, up to but not including stop, with a given step.

Here: start=0, stop=10, step=2 → [0, 2, 4, 6, 8].

# 14. What is the result of the following code?

In [14]:
import numpy as np

a = np.array([[1, 2], [3, 4]])
print(a.T)

[[1 3]
 [2 4]]


EXPLANATION:



a = np.array([[1, 2], [3, 4]])
print(a.T)

Explanation:

a is a 2×2 matrix:

[[1 2]
 [3 4]]


.T returns the transpose of the matrix, swapping rows and columns:

[[1 3]
 [2 4]]


Output:

[[1 3]
 [2 4]]

# 15. Which function finds the mean while ignoring NaN values in NumPy?

In [15]:
import numpy as np

arr = np.array([1, 2, np.nan, 4])

print(np.nanmean(arr))

2.3333333333333335


EXPLANATION:
np.nan represents a missing value.

np.nanmean(arr) calculates the mean while ignoring NaN values.

Mean of [1, 2, 4] → (1 + 2 + 4) / 3 = 2.3333

Output:

2.3333333333333335

# Q16. What does axis = 0 represent in NumPy aggregation?

In [16]:
import numpy as np

a = np.array([[1, 2, 3],
              [4, 5, 6]])

print(np.sum(a, axis=0))


[5 7 9]


EXPLANATION:
a is a 2×3 matrix:

[[1 2 3]
 [4 5 6]]


np.sum(a, axis=0) sums along the rows, i.e., sums each column:

Column 1: 1 + 4 = 5
Column 2: 2 + 5 = 7
Column 3: 3 + 6 = 9


Output:

[5 7 9]

# Question 17. What is the output shape of the following code?

## a = np.array([1,2,3])
## a.reshape(3,1)


In [17]:
import numpy as np

a = np.array([1, 2, 3])
b = a.reshape(3, 1)

print(b.shape)

(3, 1)


EXPLANATION:
a is a 1D array with shape (3,).

.reshape(3, 1) converts it into a 2D column vector with 3 rows and 1 column.

.shape returns the dimensions of the array.

Output:

(3, 1)

# Q18. What is the primary data structure in Pandas?

In [18]:
import pandas as pd

data = {"Name": ["A", "B", "C"], "Age": [20, 21, 22]}
df = pd.DataFrame(data)

print(df)

  Name  Age
0    A   20
1    B   21
2    C   22


EXPLANATION:
data is a dictionary with keys "Name" and "Age".

pd.DataFrame(data) converts the dictionary into a table-like structure (rows and columns).

Each key becomes a column, and the values become the rows.

Output:

  Name  Age
0    A   20
1    B   21
2    C   22

# Q19. What does df.head(3) return?

In [19]:
import pandas as pd

data = {"Name": ["A", "B", "C", "D", "E"], "Age": [20, 21, 22, 23, 24]}
df = pd.DataFrame(data)

df.head(3)


Name  Age
0    A   20
1    B   21
2    C   22

EXPLANATION:
df.head(n) returns the first n rows of the DataFrame.

Here, n = 3, so it shows the first 3 rows only.

Output:

  Name  Age
0    A   20
1    B   21
2    C   22

#  20. How are missing values represented in Pandas?

In [20]:
import pandas as pd
import numpy as np

data = {"A": [1, 2, np.nan, 4]}
df = pd.DataFrame(data)

print(df)


     A
0  1.0
1  2.0
2  NaN
3  4.0


EXPLANATION:
* data is a dictionary with a single key "A" and values [1, 2, np.nan, 4].

* np.nan represents a missing value.

pd.DataFrame(data) creates a DataFrame with one column "A" and four rows.

Output:

     A
0  1.0
1  2.0
2  NaN
3  4.0

# Q 21. What does df.isnull().sum() provide?

In [21]:
import pandas as pd
import numpy as np

data = {"A": [1, np.nan, 3], "B": [np.nan, 5, 6]}
df = pd.DataFrame(data)

df.isnull().sum()


A    1
B    1
dtype: int64

1. df.isnull()

Returns a DataFrame of the same shape as df with True where the value is NaN (missing), and False otherwise:

       A      B
0  False   True
1   True  False
2  False  False


2. .sum()

Summing a boolean DataFrame counts True as 1 for each column.

Counts the number of missing values per column:

A    1
B    1
dtype: int64

# Q22. What is the output of the following filtering operation?

## df[df["age"] > 25]


In [22]:
import pandas as pd

data = {"name": ["A", "B", "C", "D"], "age": [22, 28, 30, 24]}
df = pd.DataFrame(data)

df[df["age"] > 25]

name  age
1    B   28
2    C   30

EXPLANATION:
Creating the DataFrame

data is a dictionary with two columns, "A" and "B".

Some values are np.nan, which represents missing data.

The resulting DataFrame looks like this:

     A    B
0  1.0  NaN
1  NaN  5.0
2  3.0  6.0


Checking for missing values with isnull()

df.isnull() returns a boolean DataFrame where:

True indicates a missing (NaN) value

False indicates a non-missing value

       A      B
0  False   True
1   True  False
2  False  False


Counting missing values with sum()

Summing over a boolean column treats True as 1 and False as 0.

So df.isnull().sum() counts the number of missing values in each column:

A    1
B    1
dtype: int64

# Q 23. What does groupby() primarily perform?

In [23]:
import pandas as pd

data = {"Department": ["IT", "HR", "IT", "HR"],
        "Salary": [50000, 40000, 60000, 45000]}

df = pd.DataFrame(data)

df.groupby("Department")["Salary"].mean()

Department
HR    42500.0
IT    55000.0
Name: Salary, dtype: float64

EXPLANATION:
1. df.groupby("Department")

Groups the DataFrame by the values in the "Department" column.

Rows with the same department are grouped together.

Selecting "Salary" column

["Salary"] chooses the Salary column to perform operations on.

2. Calculating .mean()

Computes the average salary for each department:

IT: (50000 + 60000) / 2 = 55000

HR: (40000 + 45000) / 2 = 42500

Output:

Department
HR    42500.0
IT    55000.0
Name: Salary, dtype: float64

# Q24. What is the default join type in pd.merge()?

In [24]:
import pandas as pd

df1 = pd.DataFrame({"id": [1, 2, 3], "name": ["A", "B", "C"]})
df2 = pd.DataFrame({"id": [2, 3, 4], "age": [20, 21, 22]})

pd.merge(df1, df2, on="id")

id name  age
0   2    B   20
1   3    C   21

EXPLANATION:
1. pd.merge(df1, df2, on="id")

Combines df1 and df2 based on the common column "id".

By default, merge performs an inner join, keeping only rows with matching id values in both DataFrames.

Matching rows:

* id = 2 → present in both → merged row: id=2, name=B, age=20

* id = 3 → present in both → merged row: id=3, name=C, age=21

* id = 1 → only in df1 → excluded

* id = 4 → only in df2 → excluded

Output:

   id name  age
0   2    B   20
1   3    C   21

# Q 25. What does df.loc[0] access?

In [25]:
import pandas as pd

data = {"Name": ["A", "B", "C"], "Age": [20, 21, 22]}
df = pd.DataFrame(data)

df.loc[0]


Name     A
Age     20
Name: 0, dtype: object

EXPLANATION:
1. .loc[]

.loc is used to access rows (or columns) by label/index.

Here, 0 refers to the index label of the first row.

Selected row

Returns all column values for the row with index 0:

Name    A
Age     20
Name: 0, dtype: object